From linux-kernel-owner+chrisw=40osdl.org-S932397AbVHRS7i@vger.kernel.org Thu Aug 18 12:00:13 2005
Date:   Thu, 18 Aug 2005 20:59:37 +0200
From: Patrick McHardy <kaber@trash.net>
To: Ollie Wild <aaw@rincewind.tv>
CC: linux-kernel@vger.kernel.org, Maillist netdev <netdev@oss.sgi.com>
Subject: [IPV4]: Fix DST leak in icmp_push_reply()

Based upon a bug report and initial patch by
Ollie Wild.

Signed-off-by: Patrick McHardy <kaber@trash.net>
Signed-off-by: "David S. Miller" <davem@davemloft.net>
Signed-off-by: Chris Wright <chrisw@osdl.org>
---
 net/ipv4/icmp.c |   12 ++++++------
 1 files changed, 6 insertions(+), 6 deletions(-)

Backported to Debian's 2.4.27, 
which is needed because of IPSEC backport -- Horms

Index: linux-2.6.12.y/net/ipv4/icmp.c
===================================================================
--- a/net/ipv4/icmp.c	2005-09-01 12:04:56.000000000 +0900
+++ b/net/ipv4/icmp.c	2005-09-01 12:24:09.000000000 +0900
@@ -297,12 +297,12 @@
 {
 	struct sk_buff *skb;
 
-	ip_append_data(icmp_socket->sk, icmp_glue_bits, icmp_param,
-		       icmp_param->data_len+icmp_param->head_len,
-		       icmp_param->head_len,
-		       ipc, rt, MSG_DONTWAIT);
-
-	if ((skb = skb_peek(&icmp_socket->sk->write_queue)) != NULL) {
+	if (ip_append_data(icmp_socket->sk, icmp_glue_bits, icmp_param,
+			icmp_param->data_len+icmp_param->head_len,
+			icmp_param->head_len,
+			ipc, rt, MSG_DONTWAIT) < 0)
+		ip_flush_pending_frames(icmp_socket->sk);
+	else if ((skb = skb_peek(&icmp_socket->sk->write_queue)) != NULL) {
 		struct icmphdr *icmph = skb->h.icmph;
 		unsigned int csum = 0;
